### ISSUE TITLE
"required" error description doesn't provide context

### ISSUE THREAD

**Issue by jmaupoux** [xeipuuv/gojsonschema#187]
Hi,

Validating a json file, i don't get the context when a required field is not present. Though the error is displayed.

```go
for _, desc := range result.Errors() {
   fmt.Println(desc)
}
```

Shows
```
class: class is required
```

Should display 
```
objects.3.objects.9.class: class is required
```

The code responsible is in result.go, but i'm not sure how to handle this :
```go
	if p, ok := v.Details()["property"]; ok {
		if str, isString := p.(string); isString {
			return str
		}
	}
```

**Comment by johandorland** [comment]
The code in `result.go` is probably fine, it just returns the wrong place of where the error occurred, so the issue is probably that the created error contains the wrong context. This all happens in `validation.go`. At first glance there doesn't seem to be anything wrong with that code either. Do you have a snippet of code to reproduce this error?

**Comment by jmaupoux** [comment]
You can try the following gist

https://gist.github.com/jmaupoux/a34e574442d58112d673ff11a7082ec5

output
```
lastName: lastName is required
objects.0.age: Invalid type. Expected: integer, given: string
```

**Comment by johandorland** [comment]
Thanks for the good reproducer, that makes understanding the issue for me a whole lot easier. It seems your initial assessment was correct. According to the README:

`
err.Field(): string Returns the fieldname in the format firstName, or for embedded properties, person.firstName. This returns the same as the String() method on err.Context() but removes the (root). prefix.`

There are a couple of places that deal with the validation of properties specifically that set `"property"` in the error details, but they always only set the name of the property, not the full context. I think it's safe to remove the whole if statement and to just always return the stripped context.

Having said that I do have one little nitpick 

```
objects.3.objects.9.class: class is required
```

The object `objects.3.objects.9.class` doesn't exist. A context must always be an existing point in the JSON document that is being validated. The error therefore should be:

```
objects.3.objects.9: class is required
```

The error occurs in the object that is missing the property, not in a non-existing place. I checked other JSON schema implementations and they all raise the error in the object missing the property as well.

**Comment by johandorland** [comment]
Fixed by #218
